lib/checkleakautovar.cpp

```diff
@@ -548,6 +548,21 @@ void CheckLeakAutoVar::checkScope(const Token * const startToken,
         else if (Token::Match(tok, "continue|break ;")) {
             varInfo->clear();
         }
+
+        // Check smart pointer
+        else if (Token::Match(ftok, "%type% < %type%") && (ftok->str() == "auto_ptr" || ftok->str() == "unique_ptr" || ftok->str() == "shared_ptr")) {
+
+            const Token * endToken = ftok->linkAt(1);
+            // TODO: Check deallocator
+            bool arrayDelete = false;
+            if(Token::findsimplematch(ftok->tokAt(1), "[ ]", endToken))
+                arrayDelete = true;
+            if(!Token::Match(endToken, "> %var% {|( %var%"))
+                continue;
+            const Token * vtok = endToken->tokAt(3);
+            const VarInfo::AllocInfo allocation(arrayDelete ? -2 : -1, VarInfo::DEALLOC);
```

Comment:
you should not fix it now.. but I think this -2 / -1 hardcoding is ugly. We should add some constants. I had no idea what those meant and had to look through the code.

Reply:
I agree. I just used what was used for the `delete`. Do you know how it works? When its positive, does it represent which deallocation function in the library? As it actually needs to set the deallocation based on the custom deleter being used for the smart pointer.

Reply:
yes when it is a positive number that means some kind of library alloc. this should of course be written clearly in a comment. I will try to comment it better.

I was also thinking that:
 * it seems the next step would be to ensure the data is matching. we want a error message here:  `data = malloc(10); std::unique_ptr<T> p(data);`. If you don't add that now you could perhaps write a TODO/FIXME.
 * You've got some comments for the previous PRs. I think "the best" comments should be added as TODOs or FIXMEs in the code. Otherwise they will be forgotten. Feel free to look at that, I know I made some comment about `ValueFlow` I will see if I can add that.

Reply:
sorry I read your other comment now.. so you were already thinking about mismatching type. good.

Reply:
For your information I added the constants NEW and NEW_ARRAY now to start with.